sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroIO.java

```diff
@@ -1165,14 +1175,29 @@ public ResourceId apply(String input) {
       if (tempDirectory == null) {
         tempDirectory = getFilenamePrefix();
       }
-      WriteFiles<UserT, DestinationT, OutputT> write =
-          WriteFiles.to(
-              new AvroSink<>(tempDirectory, resolveDynamicDestinations(), getGenericRecords()));
+      final DynamicAvroDestinations<UserT, DestinationT, OutputT> dynamicDestinations =
+          resolveDynamicDestinations();
+
+      FileIO.Write<DestinationT, UserT> write =
+          FileIO.<DestinationT, UserT>writeDynamic()
+              .by(new DestinationFnViaDynamicDestinations<>(dynamicDestinations))
+              .via(
+                  new SinkFnViaDynamicDestinations<>(
+                      dynamicDestinations, getGenericRecords(), getCodec(), getMetadata()))
+              .to(
+                  new FilenamePolicyFnViaDynamicDestinations<>(
+                      getWindowedWrites(), dynamicDestinations))
+              .withDestinationCoder(
+                  dynamicDestinations.getDestinationCoderWithDefault(
+                      input.getPipeline().getCoderRegistry()))
+              .withEmptyGlobalWindowDestination(dynamicDestinations.getDefaultDestination())
+              .withTempDirectory(tempDirectory)
+              .withSideInputs(dynamicDestinations.getSideInputs());
```

Comment:
<!--new_thread; commit:28d55b3a9ab8a0cd84e7834b34c2c150edf871d7; resolved:0-->
Is this actually simpler? In particular things like getDestinationCoderWithDefault were handled automatically before, and now seems like users of FileIO need to know about this specifically.